api/v1alpha1/applicationset_types.go

```diff
@@ -93,16 +110,85 @@ type ApplicationSetBaseGenerator struct {
 	PullRequest             *PullRequestGenerator `json:"pullRequest,omitempty"`
 }
 
+type ApplicationSetTerminalGenerators []ApplicationSetTerminalGenerator
+
+// toApplicationSetNestedGenerators converts a terminal generator (a generator which cannot be a combination-type
+// generator) to a "nested" generator. The conversion is for convenience, allowing generator g to be used where a nested
+// generator is expected.
+func (g ApplicationSetTerminalGenerators) toApplicationSetNestedGenerators() []ApplicationSetNestedGenerator {
+	nestedGenerators := make([]ApplicationSetNestedGenerator, len(g))
+	for i, terminalGenerator := range g {
+		nestedGenerators[i] = ApplicationSetNestedGenerator{
+			List:                    terminalGenerator.List,
+			Clusters:                terminalGenerator.Clusters,
+			Git:                     terminalGenerator.Git,
+			SCMProvider:             terminalGenerator.SCMProvider,
+			ClusterDecisionResource: terminalGenerator.ClusterDecisionResource,
+			PullRequest:             terminalGenerator.PullRequest,
+		}
+	}
+	return nestedGenerators
+}
+
 // ListGenerator include items info
 type ListGenerator struct {
 	Elements []apiextensionsv1.JSON `json:"elements"`
 	Template ApplicationSetTemplate `json:"template,omitempty"`
 }
 
-// MatrixGenerator include Other generators
+// MatrixGenerator generates the cartesian product of two sets of parameters. The parameters are defined by two nested
+// generators.
 type MatrixGenerator struct {
-	Generators []ApplicationSetBaseGenerator `json:"generators"`
-	Template   ApplicationSetTemplate        `json:"template,omitempty"`
+	Generators []ApplicationSetNestedGenerator `json:"generators"`
+	Template   ApplicationSetTemplate          `json:"template,omitempty"`
+}
+
+// NestedMatrixGenerator is a MatrixGenerator nested under another combination-type generator (MatrixGenerator or
+// UnionGenerator). NestedMatrixGenerator does not have an override template, because template overriding has no meaning
+// within the constituent generators of combination-type generators.
+type NestedMatrixGenerator struct {
+	Generators ApplicationSetTerminalGenerators `json:"generators"`
+}
+
+// ToMatrixGenerator converts a NestedMatrixGenerator to a MatrixGenerator. This conversion is for convenience, allowing
+// a NestedMatrixGenerator to be used where a MatrixGenerator is expected (of course, the converted generator will have
+// no override template).
+func (g NestedMatrixGenerator) ToMatrixGenerator() *MatrixGenerator {
+	return &MatrixGenerator{
+		Generators: g.Generators.toApplicationSetNestedGenerators(),
+	}
+}
```

Comment:
I'm not sure I love this pattern. Would it be better to use reflection, or have `MatrixGenerator` and `NestedMatrixGenerator` implement the same interface?

Option 1 means reflection (lose some type checking at compile time).

Option 2 means defining redundant getters.

Meh, I don't love the status quo or either of the alternatives.

Reply:
Personally I would go with defining a common interface which brings other advantages like more flexible code and easier to test/mock. My second choice would be having the conversion functions like the way you implemented which seems fine to me as well.
The solution I would really avoid is going with reflection as this is such a simple problem to handle and there are many negative side effects in going with reflection like:
- lower maintainability
- higher complexity
- harder to test
- losing compile time type checking like you described

Reply:
I need to take a few hours one day and attempt the interface solution to see how it feels.

I think the good news is that it's completely an implementation detail which could be applied later. (Let me know if that assumption might be incorrect. If that's the case, we can evaluate the interface approach before merging this.)